kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,991 @@
+    /// - `Err(ErrorCode)`: If there is nowhere to store the process a suitable
+    ///   `ErrorCode` will be returned.
+    fn setup(&self, app_length: usize) -> Result<usize, ErrorCode>;
+
+    /// Instruct the kernel to write data to the flash
+    ///
+    ///This is used to write both userland apps and padding apps
+    fn write_app_data(
+        &self,
+        buffer: &'static mut [u8],
+        offset: usize,
+        app_size: usize,
+        processid: ProcessId,
+    ) -> Result<(), ErrorCode>;
+
+    /// Instruct the kernel to load and execute the process.
+    ///
+    /// This tells the kernel to try to actually execute the new process that
+    /// was just installed in flash from the preceding `write()` call.
+    fn load(&self) -> Result<(), ErrorCode>;
+
+    /// Sets a client for the DynamicProcessLoading Object
+    ///
+    /// When the client operation is done, it calls the
+    /// app_data_write_done(&self, buffer: &'static mut [u8], length: usize) function
+    fn set_client(&self, client: &'static dyn DynamicProcessLoadingClient);
+}
+
+/// The callback for set_client(&self, Client).
+/// The client capsule should implement this trait to handle the callback logic
+pub trait DynamicProcessLoadingClient {
+    fn app_data_write_done(&self, buffer: &'static mut [u8], length: usize);
+}
+
+pub struct DynamicProcessLoader<'a, C: 'static + Chip> {
+    kernel: &'static Kernel,
+    chip: &'static C,
+    fault_policy: &'static dyn ProcessFaultPolicy,
+    procs: MapCell<&'static mut [Option<&'static dyn process::Process>]>,
+    flash: Cell<&'static [u8]>,
+    app_memory: OptionalCell<&'static mut [u8]>,
+    new_process_flash: OptionalCell<&'static [u8]>,
+    flash_driver: &'a dyn NonvolatileStorage<'a>,
+    buffer: TakeCell<'static, [u8]>,
+    new_app_start_addr: Cell<usize>,
+    new_app_length: Cell<usize>,
+    client: OptionalCell<&'static dyn DynamicProcessLoadingClient>,
+    state: OptionalCell<State>,
+}
+
+impl<'a, C: 'static + Chip> DynamicProcessLoader<'a, C> {
+    pub fn new(
+        processes: &'static mut [Option<&'static dyn process::Process>],
+        kernel: &'static Kernel,
+        chip: &'static C,
+        flash: &'static [u8],
+        fault_policy: &'static dyn ProcessFaultPolicy,
+        driver: &'a dyn NonvolatileStorage<'a>,
+        buffer: &'static mut [u8],
+    ) -> Self {
+        Self {
+            procs: MapCell::new(processes),
+            kernel,
+            chip,
+            flash: Cell::new(flash),
+            app_memory: OptionalCell::empty(),
+            fault_policy,
+            new_process_flash: OptionalCell::empty(),
+            flash_driver: driver,
+            buffer: TakeCell::new(buffer),
+            new_app_start_addr: Cell::new(0),
+            new_app_length: Cell::new(0),
+            client: OptionalCell::empty(),
+            state: OptionalCell::empty(),
+        }
+    }
+
+    // Needs to be set separately, or breaks grants allocation
```

Comment:
While I agree this needs to be a function, what does breaking grant allocation mean?

Reply:
Explained better.